File: 2023/day-3/part-1.cs
using System;
using System.IO;

string inputFile = @"./input.txt";
if (!File.Exists(inputFile))
{
    Console.WriteLine($"Couldn't find file named: {inputFile.TrimStart("./".ToCharArray())}");
    return;

}

int sum = 0;
string[] lines = File.ReadAllLines(inputFile);

for (int i = 0; i < lines.Length; i++)
{
    string currentline = lines[i];
    string prevLine = "";
    string nextLine = "";

    if (i - 1 >= 0)
    {
        prevLine = lines[i - 1];
    }
    if (i + 1 < lines.Length)
    {
        nextLine = lines[i + 1];
    }

    string partNumber = "";
    bool isPartNumberValid = false;
    for (int j = 0; j < currentline.Length; j++)
    {
        if (char.IsDigit(currentline[j]))
        {
            partNumber += currentline[j];

            if (isPartNumberValid)
            {
                continue;
            }

            char[] diagonalCharacters = new char[8];

            diagonalCharacters[0] = currentline.ElementAtOrDefault(j - 1);
            diagonalCharacters[1] = currentline.ElementAtOrDefault(j + 1);

            diagonalCharacters[2] = prevLine.ElementAtOrDefault(j);
            diagonalCharacters[3] = prevLine.ElementAtOrDefault(j - 1);
            diagonalCharacters[4] = prevLine.ElementAtOrDefault(j + 1);

            diagonalCharacters[5] = nextLine.ElementAtOrDefault(j);
            diagonalCharacters[6] = nextLine.ElementAtOrDefault(j - 1);
            diagonalCharacters[7] = nextLine.ElementAtOrDefault(j + 1);

            for (int k = 0; k < 8; k++)
            {
                if (isSymbol(diagonalCharacters[k]))
                {
                    Console.WriteLine($"Part Number: {partNumber} and Symbol at {k}: {diagonalCharacters[k]} \nPrevLine: {prevLine} \nNextLine: {nextLine}\n");
                    isPartNumberValid = true;
                    break;
                }
            }
        }
        else if (currentline[j] == '.')
        {
            int num;
            if ((isPartNumberValid) && (Int32.TryParse(partNumber, out num)))
            {
                // Console.WriteLine($"Part Number: {partNumber}");
                sum += num;
                isPartNumberValid = false;
            }
            partNumber = "";
        }

    }

}

Console.WriteLine($"Sum: {sum}");

bool isSymbol(char ch)
{
    string symbols = @"/@#!$%^&*()+-=";
    return symbols.Contains(ch);
}

File: 2023/day-3/part1/part-1.py
def examine_adjacent_characters(char_index, adjacent_lines):
    for line in adjacent_lines:
        if line[char_index - 1] in SYMBOLS or line[char_index] in SYMBOLS or line[char_index + 1] in SYMBOLS:
            return True
    return False

def calculate_sum_of_part_numbers(lines):
    sum = 0
    for line_index in range(len(lines)):
        current_line = lines[line_index]
        prev_line = lines[line_index - 1] if line_index != 0 else ["." for i in range(len(current_line))]
        next_line = lines[line_index + 1] if line_index + 1 < len(lines) else ["." for i in range(len(current_line))]

        adjacent_lines = [current_line, prev_line, next_line]

        partNumber = ""
        isValidPartNumber = False

        for char_index in range(len(current_line)):
            char = current_line[char_index]

            if char.isdigit():
                partNumber += char
                isValidPartNumber = isValidPartNumber or examine_adjacent_characters(char_index, adjacent_lines)

            if char == '.' or char in SYMBOLS:
                if isValidPartNumber:
                    sum += int(partNumber)
                isValidPartNumber = False
                partNumber = ""

            if char_index + 2 == len(current_line):
                if current_line[char_index + 1].isdigit():
                    partNumber += current_line[char_index + 1]
                    if isValidPartNumber:
                        sum += int(partNumber)
                    break
    return sum

SYMBOLS = "!@#$%^&*()_+-=|<>,:;/"
with open("../input.txt") as file:
    lines = file.read().split("\n")
    print(f"Sum: {calculate_sum_of_part_numbers(lines)}")